src/features/unified-share-modal/UnifiedShareForm.js

```diff
@@ -54,15 +69,70 @@ class UnifiedShareForm extends React.Component<USFProps, State> {
         super(props);
 
         this.state = {
+            classificationLabelId: '',
             emailSharedLinkContacts: [],
             inviteCollabsContacts: props.initiallySelectedContacts,
             inviteePermissionLevel: '',
             isEmailLinkSectionExpanded: false,
+            isFetchingJustificationReasons: false,
             isInviteSectionExpanded: !!props.initiallySelectedContacts.length,
+            justificationReasons: [],
             showCollaboratorList: false,
         };
     }
 
+    componentDidUpdate(prevProps: USFProps) {
+        const { isCollabRestrictionJustificationAllowed, item, restrictedExternalCollabEmails } = this.props;
+        const {
+            restrictedExternalCollabEmails: prevRestrictedExternalCollabEmails,
+            isCollabRestrictionJustificationAllowed: prevIsCollabRestrictionJustificationAllowed,
+        } = prevProps;
+
+        const didExternalCollabRestrictionsChange =
+            !isEqual(restrictedExternalCollabEmails, prevRestrictedExternalCollabEmails) ||
+            isCollabRestrictionJustificationAllowed !== prevIsCollabRestrictionJustificationAllowed;
+
+        if (didExternalCollabRestrictionsChange && this.shouldRequireExternalCollabJustification()) {
+            this.fetchJustificationReasons(item, JUSTIFICATION_CHECKPOINT_EXTERNAL_COLLAB);
+        }
+    }
+
+    fetchJustificationReasons = (item: Item, checkpoint: JustificationCheckpoint) => {
+        const { justificationReasons } = this.state;
+        const { getJustificationReasons } = this.props;
+        const hasJustificationReasons = !!justificationReasons.length;
+
+        if (!getJustificationReasons || hasJustificationReasons) {
+            return Promise.resolve();
+        }
+        this.setState({ isFetchingJustificationReasons: true });
+
+        return getJustificationReasons(item.typedID, checkpoint)
+            .then(({ classificationLabelId, options }: GetJustificationReasonsResponse) => {
```

Comment:
Using promises instead of `async`/`await` for consistency with the rest of the file.